pkg/releaselock/controller.go

```diff
@@ -0,0 +1,203 @@
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rpc
+
+import (
+	"context"
+	"os"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apiError "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/gcp-filestore-csi-driver/pkg/util"
+
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+)
+
+const (
+	gceInstanceIDKey = "container.googleapis.com/instance_id"
+	leaseName        = "filestore-csi-storage-gke-io-node"
+	// Root CA configmap in each namespace.
+	rootCA = "kube-root-ca.crt"
+
+	defaultLeaderElection = true
+	defaultLeaseDuration  = 15 * time.Second
+	defaultRenewDeadline  = 10 * time.Second
+	defaultRetryPeriod    = 2 * time.Second
+	defaultSyncPeriod     = 3 * time.Minute
+)
+
+type LockReleaseController struct {
+	client kubernetes.Interface
+
+	// Identity of this controller, generated at creation time and not persisted
+	// across restarts. Useful only for debugging, for seeing the source of events.
+	id string
+
+	// Whether to do kubernetes leader election at all. It should basically
+	// always be done when possible to avoid duplicate Provision attempts.
+	leaderElection          bool
+	leaderElectionNamespace string
+	// Parameters of leaderelection.LeaderElectionConfig.
+	leaseDuration, renewDeadline, retryPeriod time.Duration
+
+	// Reconcile loop frequency.
+	syncPeriod time.Duration
+}
+
+func NewLockReleaseController(client kubernetes.Interface) (*LockReleaseController, error) {
+	// Register rpc procedure for lock release.
+	if err := RegisterLockReleaseProcedure(); err != nil {
+		klog.Errorf("Error initializing lockrelease controller: %v", err)
+		return nil, err
+	}
+
+	id, err := os.Hostname()
+	if err != nil {
+		klog.Errorf("Failed to get hostname for lockrelease controller: %v", err)
+		return nil, err
+	}
+	// Add a uniquifier so that two processes on the same host don't accidentally both become active.
+	id = id + "_" + string(uuid.NewUUID())
+
+	return &LockReleaseController{
+		id:                      id,
+		client:                  client,
+		leaderElection:          defaultLeaderElection,
+		leaderElectionNamespace: util.ConfigMapNamespace,
+		leaseDuration:           defaultLeaseDuration,
+		renewDeadline:           defaultRenewDeadline,
+		retryPeriod:             defaultRetryPeriod,
+		syncPeriod:              defaultSyncPeriod,
+	}, nil
+}
+
+func (c *LockReleaseController) Run(ctx context.Context) {
+	run := func(ctx context.Context) {
+		klog.Infof("Lock release controller %s started leading", c.id)
```

Comment:
started on node %s.

this helps grep across all nodes to trace the leadership changes.

Reply:
done